File: frontend/src/types/post.ts
// Post and Storage Types

import type { SlideDesign } from './slide';
import type { LayoutConfig } from './template';

export interface StorageUrls {
  slides: string[];
  thumbnail: string | null;
}

export interface PostAnalytics {
  impressions: number;
  engagement: number;
  saves: number;
  shares: number;
  engagementRate: number;
  lastUpdated: Date | null;
}

export interface PostMetadata {
  variantLabel: string | null;
  generationParams: Record<string, any>;
}

/**
 * Slide in a post - derived from template's SlideDesign
 * with dynamic content filled in by AI
 */
export interface PostSlide extends Omit<SlideDesign, 'name'> {
  id: string;
  slideNumber: number;
  designId: string; // Reference to original SlideDesign.id from template
  imagePrompt?: string; // Optional AI-generated background prompt
}

export interface PostContent{
  slides: PostSlide[];
  layout: LayoutConfig;

  caption: string;
  hashtags: string[];
}

export type PostStatus = 'draft' | 'scheduled' | 'published' | 'failed';
export type Platform = 'instagram' | 'tiktok';

export interface Post {
  id: string;
  userId: string;
  templateId: string;
  variantSetId?: string;
  platform: Platform;
  status: PostStatus;
  createdAt: Date;
  updatedAt?: Date;
  scheduledTime?: Date;
  publishedTime?: Date;
  content: PostContent;
  storageUrls: StorageUrls;
  analytics: PostAnalytics;
  metadata: PostMetadata;
}

export interface CreatePostRequest {
  templateId: string;
  platform: Platform;
  content: PostContent;
  scheduledTime?: Date;
  variantSetId?: string;
}

export interface UpdatePostRequest {
  status?: PostStatus;
  scheduledTime?: Date;
  content?: PostContent;
}

// Extended Post with template info
export interface PostWithTemplate extends Post {
  template?: {
    id: string;
    name: string;
    category: string;
  };
  analyticsHistory?: any[];
}

File: backend/models/post.py
# Post Models

from pydantic import BaseModel, Field
from typing import List, Optional, Literal
from datetime import datetime
from .gemini import GeminiCarouselResponse

# ==================== POST SUB-MODELS ====================

class StorageUrls(BaseModel):
    slides: List[str] = []  # URLs to slide images in Supabase Storage
    thumbnail: Optional[str] = None  # URL to thumbnail image

class PostAnalytics(BaseModel):
    impressions: int = 0
    engagement: int = 0
    saves: int = 0
    shares: int = 0
    engagementRate: float = 0.0
    lastUpdated: Optional[datetime] = None

class PostMetadata(BaseModel):
    variantLabel: Optional[str] = None
    generationParams: dict = {}

# ==================== POST MODELS ====================

class Post(BaseModel):
    id: str
    user_id: str
    template_id: str
    variant_set_id: Optional[str] = None
    platform: str  # Supabase: string
    status: str  # Supabase: string
    created_at: str  # Supabase: string (ISO)
    updated_at: Optional[str] = None
    scheduled_time: Optional[str] = None
    published_time: Optional[str] = None
    content: dict  # Supabase: JSON
    storage_urls: dict = {}
    metadata: dict = {}

class CreatePostRequest(BaseModel):
    templateId: str
    platform: Literal["instagram", "tiktok"]
    content: GeminiCarouselResponse
    scheduledTime: Optional[datetime] = None
    variantSetId: Optional[str] = None

class UpdatePostRequest(BaseModel):
    status: Optional[Literal["draft", "scheduled", "published", "failed"]] = None
    scheduledTime: Optional[datetime] = None
    content: Optional[GeminiCarouselResponse] = None
